File: bench/algorithm/fft/1.ts
import {Complex} from "https://deno.land/x/complex/mod.ts";

const PI = Math.PI;

// Recursive FFT implementation
function fft(arr: Complex[], invert: boolean): Complex[] {
  const n = arr.length;
  if (n === 1) {
    return arr;
  }
  const even = new Array(n / 2).fill(0).map((_, i) => arr[2 * i]);
  const odd = new Array(n / 2).fill(0).map((_, i) => arr[2 * i + 1]);
  const fftEven = fft(even, invert);
  const fftOdd = fft(odd, invert);
  const result = new Array(n).fill(0);
  for (let i = 0; i < n / 2; i++) {
    const angle = 2 * PI * i / n;
    const w = invert ? new Complex(Math.cos(angle), -Math.sin(angle)) : new Complex(Math.cos(angle), Math.sin(angle));
    result[i] = fftEven[i].add(w.mul(fftOdd[i]));
    result[i + n / 2] = fftEven[i].sub(w.mul(fftOdd[i]));
  }
  return result;
}

// Main FFT function
export function fftTransform(arr: number[]): Complex[] {
  const n = arr.length;
  const result = new Array(n).fill(0).map((_, i) => new Complex(arr[i], 0));
  return fft(result, false);
}

// Inverse FFT function
export function ifftTransform(arr: Complex[]): Complex[] {
  const n = arr.length;
  const result = fft(arr, true);
  return new Array(n).fill(0).map((_, i) => result[i].div(n));
}

File: bench/algorithm/fft/1.go
package main

import (
	"fmt"
	"math"
	"math/cmplx"
)

func fft(x []complex128) []complex128 {
	N := len(x)
	if N == 1 {
		return x
	}
	X := make([]complex128, N)
	for k := 0; k < N/2; k++ {
		X[k] = x[2*k] + cmplx.Exp(complex(0, -2*math.Pi*float64(k)/float64(N)))*x[2*k+1]
		X[k+N/2] = x[2*k] - cmplx.Exp(complex(0, -2*math.Pi*float64(k)/float64(N)))*x[2*k+1]
	}
	return X
}

func main() {
	x := make([]complex128, 4)
	for i := 0; i < 4; i++ {
		x[i] = complex(float64(i), 0)
	}
	fmt.Println(fft(x))
}
